packages/bp/src/core/nlu/nlu-inference-service.ts

```diff
@@ -0,0 +1,161 @@
+const EVENTS_TO_IGNORE = ['session_reference', 'session_reset', 'bp_dialog_timeout', 'visit', 'say_something', '']
+const PREDICT_MW = 'nlu-predict.incoming'
+
+/**
+ * This service takes care of the nlu inferences/predictions
+ *
+ * It belongs in the runtime and only in the runtime.
+ * It is not responsible for CRUD on intent and entity files.
+ * It is not responsible for trainings and the models state-machine.
+ * It expects bot config to have field nluModels.
+ * It expects botpress config to have field nlu.
+ * That's it.
+ */
+@injectable()
+export class NLUInferenceService {
+  private legacyElection!: boolean
+
+  private predictors: { [botId: string]: Predictor } = {}
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.ConfigProvider) private configProvider: ConfigProvider,
+    @inject(TYPES.Logger) private logger: Logger,
+    @inject(TYPES.NLUClientProvider) private nluClientProvider: NLUClientProvider
+  ) {}
+
+  @postConstruct()
+  public async initialize() {
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
+    await this.nluClientProvider.initialize()
+    const { nlu: nluConfig } = await this.configProvider.getBotpressConfig()
+
+    const { legacyElection } = nluConfig
+    this.legacyElection = legacyElection ?? false
+
+    this.eventEngine.register({
+      name: PREDICT_MW,
+      description:
+        'Process natural language in the form of text. Structured data with an action and parameters for that action is injected in the incoming message event.',
+      order: 100,
+      direction: 'incoming',
+      handler: this.handleIncomingEvent.bind(this)
+    })
+  }
+
+  public async teardown() {
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
+    this.eventEngine.removeMiddleware(PREDICT_MW)
+  }
+
+  public async mountBot(botId: string) {
+    await AppLifecycle.waitFor(AppLifecycleEvents.STUDIO_READY)
+    await this.nluClientProvider.mountBot(botId)
+    const botConfig = await this.configProvider.getBotConfig(botId)
+    const modelIdGetter = this._modelIdGetter(botId)
+
+    const client = this.nluClientProvider.getClientForBot(botId)!
+    const predictor = new Predictor(botId, botConfig.defaultLanguage, client, modelIdGetter, this.logger)
+    this.predictors[botId] = predictor
+  }
+
+  public async unmountBot(botId: string) {
+    this.nluClientProvider.unmountBot(botId)
+    delete this.predictors[botId]
+  }
+
+  private _modelIdGetter = (botId: string) => async (): Promise<Dic<string>> => {
+    // TODO: implement some caching to prevent from reading bot config at each predict
+    const botConfig = await this.configProvider.getBotConfig(botId)
+    return botConfig.nluModels ?? {}
+  }
+
+  private handleIncomingEvent = async (event: IO.Event, next: IO.MiddlewareNextCallback) => {
+    if (!this._isIncoming(event)) {
+      const err = new Error('NLU incoming mw should receive incoming event.')
+      return next(err)
+    }
```

Comment:
This check is not necessary, you already register it as an incoming handler... This case will never happen

Reply:
this is because typings are shit: Incoming middlewares should receive incoming events 

Anyway, I'll use a cast instead (`event as IncomingEvent`).